internal/controllers/core/cluster/reconciler.go

```diff
@@ -104,7 +109,7 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 }
 
 // Creates a docker connection from the spec.
-func (r *Reconciler) createDockerConnection(ctx context.Context, obj *v1alpha1.DockerClusterConnection) *connection {
+func (r *Reconciler) createDockerConnection(ctx context.Context, obj *v1alpha1.DockerClusterConnection) connection {
```

Comment:
Curious about Go type signature design here -- what led you to change this from `*connection` to `connection` (and corresponding cascading changes)? The connection struct is big enough that it would seem like passing copies of it around would be a little more expensive -- aren't you duplicating all the possible state in the k8s/docker client structs as well?

Reply:
I switched to value types for `connection` to avoid goroutine-safety pitfalls - I think a lot of the Tilt codebase has used this pattern to good effect where we copy a value type, mutate, and then wholesale replace the entry in the shared map, so that we don't have to worry it.

From a performance perspective, the client types are interfaces, so they're both pointers and we won't actually copy the full structs. The prevailing wisdom I've seen for Go code is that preferring value types where appropriate is the right place to start and then change if it presents a performance problem. Copying the value and keeping it on stack is often actually much faster than having stuff on heap!

Reply:
Thanks! That helps a lot.